README.md
# eth-account

## Quickstart

```sh
python -m pip install eth-account
```

## Developer Setup

for information on how we do:

-   Testing
-   Pull Requests
-   Code Style
-   Documentation

### Development Environment Setup

You can set up your dev environment with:

```sh
git clone git@github.com:ethereum/eth-account.git
cd eth-account
virtualenv -p python3 venv
. venv/bin/activate
python -m pip install -e ".[dev]"
```

To run the integration test cases, you need to install node and the custom cli tool as follows:

```sh
apt-get install -y nodejs  # As sudo
./tests/integration/ethers-cli/setup_node_v18.sh  # As sudo
cd tests/integration/ethers-cli
npm install -g .  # As sudo
```

### Release setup

To release a new version:

```sh
make release bump=$$VERSION_PART_TO_BUMP$$
```

#### How to bumpversion

The version format for this repo is `{major}.{minor}.{patch}` for stable, and
`{major}.{minor}.{patch}-{stage}.{devnum}` for unstable (`stage` can be alpha or beta).

To issue the next version in line, specify which part to bump,
like `make release bump=minor` or `make release bump=devnum`. This is typically done from the
master branch, except when releasing a beta (in which case the beta is released from master,
and the previous stable branch is released from said branch).

If you are in a beta version, `make release bump=stage` will switch to a stable.

To issue an unstable version when the current version is stable, specify the
new version explicitly, like `make release bump="--new-version 4.0.0-alpha.1 devnum"`


docs/release_notes.rst
Release Notes
=============

.. towncrier release notes start

eth-account v0.9.0 (2023-06-07)
-------------------------------

Breaking Changes
~~~~~~~~~~~~~~~~

- drop python3.6 support from setup (`#228 <https://github.com/ethereum/eth-account/issues/228>`__)


Improved Documentation
~~~~~~~~~~~~~~~~~~~~~~

- remove notices of Draft status for eips 712 and 191 (`#222 <https://github.com/ethereum/eth-account/issues/222>`__)


Features
~~~~~~~~

- Add support for Python 3.11 (`#212 <https://github.com/ethereum/eth-account/issues/212>`__)


Internal Changes - for eth-account Contributors
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- Upgrade Node from v12.x to v18.x in tests (`#217 <https://github.com/ethereum/eth-account/issues/217>`__)
- pulled full new node_v18 install script (`#223 <https://github.com/ethereum/eth-account/issues/223>`__)
- bump versions for docs dependencies (`#224 <https://github.com/ethereum/eth-account/issues/224>`__)
- add sphinx_rtd_theme to docs/conf.py extensions list (`#225 <https://github.com/ethereum/eth-account/issues/225>`__)
- merge in updates from python project template (`#288 <https://github.com/ethereum/eth-account/issues/288>`__)


eth-account v0.8.0 (2022-12-15)
-------------------------------

Features
~~~~~~~~

- update all references to deprecated `eth_abi.encode_abi` to `eth_abi.encode` (`#200 <https://github.com/ethereum/eth-account/issues/200>`__)


Performance improvements
~~~~~~~~~~~~~~~~~~~~~~~~

- Reduce the number of pbkdf2 iterations to speed up tests (`#77 <https://github.com/ethereum/eth-account/issues/77>`__)


Deprecations and Removals
~~~~~~~~~~~~~~~~~~~~~~~~~

- remove deprecated methods that were noted to go in v0.5 (`#195 <https://github.com/ethereum/eth-account/issues/195>`__)


Internal Changes - for eth-account Contributors
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- add coverage reporting to pytest (`#192 <https://github.com/ethereum/eth-account/issues/192>`__)
- Use updated circleci Python images, fix Sphinx warning (`#194 <https://github.com/ethereum/eth-account/issues/194>`__)


Miscellaneous changes
~~~~~~~~~~~~~~~~~~~~~

- `#197 <https://github.com/ethereum/eth-account/issues/197>`__, `#198 <https://github.com/ethereum/eth-account/issues/198>`__, `#199 <https://github.com/ethereum/eth-account/issues/199>`__, `#202 <https://github.com/ethereum/eth-account/issues/202>`__, `#203 <https://github.com/ethereum/eth-account/issues/203>`__, `#204 <https://github.com/ethereum/eth-account/issues/204>`__, `#206 <https://github.com/ethereum/eth-account/issues/206>`__


eth-account v0.7.0 (2022-08-17)
-------------------------------

Bugfixes
~~~~~~~~

- bump ansi-regex to 5.0.1 to fix minor ReDos vulnerability (`#129 <https://github.com/ethereum/eth-account/issues/129>`__)
- Enable lint runs again on CI (`#166 <https://github.com/ethereum/eth-account/issues/166>`__)
- fix DoS-able regex pattern (`#178 <https://github.com/ethereum/eth-account/issues/178>`__)
- Allow towncrier to build the release notes again (`#185 <https://github.com/ethereum/eth-account/issues/185>`__)


Improved Documentation
~~~~~~~~~~~~~~~~~~~~~~

- Add example to generate multiple accounts from a mnemonic (`#153 <https://github.com/ethereum/eth-account/issues/153>`__)
- Pin Jinja2 at >=3.0.0,<3.1.0; pin towncrier==18.5.0; open up Sphinx requirement to allow >=1.6.5,<5. (`#156 <https://github.com/ethereum/eth-account/issues/156>`__)
- added missing quotes to readme dev environment setup example (`#172 <https://github.com/ethereum/eth-account/issues/172>`__)


Miscellaneous changes
~~~~~~~~~~~~~~~~~~~~~

- `#79 <https://github.com/ethereum/eth-account/issues/79>`__, `#155 <https://github.com/ethereum/eth-account/issues/155>`__, `#162 <https://github.com/ethereum/eth-account/issues/162>`__, `#164 <https://github.com/ethereum/eth-account/issues/164>`__, `#165 <https://github.com/ethereum/eth-account/issues/165>`__


Breaking changes
~~~~~~~~~~~~~~~~

- Change bitarray dependency requirement to be >=2.4,<3 since 2.4 has wheels for all platform types. (`#154 <https://github.com/ethereum/eth-account/issues/154>`__)
- Fix errors in EIP-712 signing (`#175 <https://github.com/ethereum/eth-account/issues/175>`__)


eth-account v0.6.1 (2022-02-24)
-------------------------------

Bugfixes
~~~~~~~~

- Allow encoding of structured data containing ``bytes`` (`#91 <https://github.com/ethereum/eth-account/issues/91>`__)


Miscellaneous changes
~~~~~~~~~~~~~~~~~~~~~

- `#68 <https://github.com/ethereum/eth-account/issues/68>`__, `#144 <https://github.com/ethereum/eth-account/issues/144>`__


eth-account v0.6.0 (2022-01-20)
-------------------------------

Features
~~~~~~~~

- Update dependencies:
  - eth-abi
  - eth-keyfile
  - eth-keys
  - eth-rlp
  - pyrlp
  - eth-utils (`#138 <https://github.com/ethereum/eth-account/issues/138>`__)
- Add support for Python 3.9 and 3.10 (`#139 <https://github.com/ethereum/eth-account/issues/139>`__)


Deprecations and Removals
~~~~~~~~~~~~~~~~~~~~~~~~~

- Drop support for Python 3.6 (`#139 <https://github.com/ethereum/eth-account/issues/139>`__)


eth-account v0.5.9 (2022-08-04)
-------------------------------

Bugfixes
~~~~~~~~

- fix DoS-able regex pattern (`#178 <https://github.com/ethereum/eth-account/issues/178>`__)


Miscellaneous changes
~~~~~~~~~~~~~~~~~~~~~

- `#183 <https://github.com/ethereum/eth-account/issues/183>`__, `#184 <https://github.com/ethereum/eth-account/issues/184>`__


eth-account v0.5.8 (2022-06-06)
-------------------------------

Miscellaneous changes
~~~~~~~~~~~~~~~~~~~~~

- `#163 <https://github.com/ethereum/eth-account/issues/163>`__, `#168 <https://github.com/ethereum/eth-account/issues/168>`__

eth-account v0.5.7 (2022-01-27)
-------------------------------

Features
~~~~~~~~


eth_account/messages.py
from collections.abc import (
    Mapping,
)
from typing import (
    NamedTuple,
    Union,
)

from eth_typing import (
    Address,
    Hash32,
)
from eth_utils.curried import (
    ValidationError,
    keccak,
    text_if_str,
    to_bytes,
    to_canonical_address,
    to_text,
)
from hexbytes import (
    HexBytes,
)

from eth_account._utils.structured_data.hashing import (
    hash_domain,
    hash_message as hash_eip712_message,
    load_and_validate_structured_message,
)
from eth_account._utils.structured_data.validation import (
    validate_structured_data,
)
from eth_account._utils.validation import (
    is_valid_address,
)

text_to_bytes = text_if_str(to_bytes)


# watch for updates to signature format
class SignableMessage(NamedTuple):
    """
    A message compatible with EIP-191_ that is ready to be signed.

    The properties are components of an EIP-191_ signable message. Other message formats
    can be encoded into this format for easy signing. This data structure doesn't need
    to know about the original message format. For example, you can think of
    EIP-712 as compiling down to an EIP-191 message.

    In typical usage, you should never need to create these by hand. Instead, use
    one of the available encode_* methods in this module, like:

        - :meth:`encode_structured_data`
        - :meth:`encode_intended_validator`
        - :meth:`encode_defunct`

    .. _EIP-191: https://eips.ethereum.org/EIPS/eip-191
    """

    version: bytes  # must be length 1
    header: bytes  # aka "version specific data"
    body: bytes  # aka "data to sign"


def _hash_eip191_message(signable_message: SignableMessage) -> Hash32:
    version = signable_message.version
    if len(version) != 1:
        raise ValidationError(
            f"The supplied message version is {version!r}. "
            "The EIP-191 signable message standard only supports one-byte versions."
        )

    joined = b"\x19" + version + signable_message.header + signable_message.body
    return Hash32(keccak(joined))


# watch for updates to signature format
def encode_intended_validator(
    validator_address: Union[Address, str],
    primitive: bytes = None,
    *,
    hexstr: str = None,
    text: str = None,
) -> SignableMessage:
    """
    Encode a message using the "intended validator" approach (ie~ version 0)
    defined in EIP-191_.

    Supply the message as exactly one of these three arguments:
    bytes as a primitive, a hex string, or a unicode string.

    .. WARNING:: Note that this code has not gone through an external audit.

    :param validator_address: which on-chain contract is capable of validating this
        message, provided as a checksummed address or in native bytes.
    :param primitive: the binary message to be signed
    :type primitive: bytes or int
    :param str hexstr: the message encoded as hex
    :param str text: the message as a series of unicode characters (a normal Py3 str)
    :returns: The EIP-191 encoded message, ready for signing

    .. _EIP-191: https://eips.ethereum.org/EIPS/eip-191
    """
    if not is_valid_address(validator_address):
        raise ValidationError(
            f"Cannot encode message with 'Validator Address': {validator_address!r}. "
            "It must be a checksum address, or an address converted to bytes."
        )
    # The validator_address is a str or Address (which is a subtype of bytes). Both of
    # these are AnyStr, which includes str and bytes.
    # Not sure why mypy complains here...
    canonical_address = to_canonical_address(validator_address)
    message_bytes = to_bytes(primitive, hexstr=hexstr, text=text)
    return SignableMessage(
        HexBytes(b"\x00"),  # version 0, as defined in EIP-191
        canonical_address,
        message_bytes,
    )


def encode_structured_data(
    primitive: Union[bytes, int, Mapping] = None,
    *,
    hexstr: str = None,
    text: str = None,
) -> SignableMessage:
    r"""
    Encode an EIP-712_ message.

    EIP-712 is the "structured data" approach (ie~ version 1 of an EIP-191 message).

    Supply the message as exactly one of the three arguments:

        - primitive, as a dict that defines the structured data
        - primitive, as bytes
        - text, as a json-encoded string
        - hexstr, as a hex-encoded (json-encoded) string

    .. WARNING:: Note that this code has not gone through an external audit, and
        the test cases are incomplete.

    :param primitive: the binary message to be signed
    :type primitive: bytes or int or Mapping (eg~ dict )
    :param hexstr: the message encoded as hex
    :param text: the message as a series of unicode characters (a normal Py3 str)
    :returns: The EIP-191 encoded message, ready for signing


    Usage Notes:
     - An EIP712 message consists of 4 top-level keys: ``types``, ``primaryType``,
       ``domain``, and ``message``. All 4 must be present to encode properly.
     - The key ``EIP712Domain`` must be present within ``types``.
     - The `type` of a field may be a Solidity type or a `custom` type, i.e., one
       that is defined within the ``types`` section of the typed data.
     - Extra information in ``message`` and ``domain`` will be ignored when encoded.
       For example, if the custom type ``Person`` defines the fields ``name`` and
       ``wallet``, but an additional ``id`` field is provided in ``message``, the
       resulting encoding will be the same as if the ``id`` information was not present.
     - Unused custom types will be ignored in the same way.

    .. doctest:: python

        >>> # an example of basic usage
        >>> import json
        >>> from eth_account import Account
        >>> from eth_account.messages import encode_structured_data

        >>> typed_data = {
        ...     "types": {
        ...         "EIP712Domain": [
        ...             {"name": "name", "type": "string"},
        ...             {"name": "version", "type": "string"},
        ...             {"name": "chainId", "type": "uint256"},
        ...             {"name": "verifyingContract", "type": "address"},
        ...         ],
        ...         "Person": [
        ...             {"name": "name", "type": "string"},
        ...             {"name": "wallet", "type": "address"},
        ...         ],
        ...         "Mail": [
        ...             {"name": "from", "type": "Person"},
        ...             {"name": "to", "type": "Person"},
        ...             {"name": "contents", "type": "string"},
        ...         ],
        ...     },
        ...     "primaryType": "Mail",
        ...     "domain": {
        ...         "name": "Ether Mail",
        ...         "version": "1",
        ...         "chainId": 1,
        ...         "verifyingContract": "0xCcCCccccCCCCcCCCCCCcCcCccCcCCCcCcccccccC",
        ...     },
        ...     "message": {
        ...         "from": {
        ...             "name": "Cow",
        ...             "wallet": "0xCD2a3d9F938E13CD947Ec05AbC7FE734Df8DD826"
        ...         },
        ...         "to": {
        ...             "name": "Bob",
        ...             "wallet": "0xbBbBBBBbbBBBbbbBbbBbbbbBBbBbbbbBbBbbBBbB"
        ...         },
        ...         "contents": "Hello, Bob!",
        ...     },
        ... }

        >>> key = "0xaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"

        >>> signable_msg_from_dict = encode_structured_data(typed_data)
        >>> signable_msg_from_str = encode_structured_data(text=json.dumps(typed_data))
        >>> signable_msg_from_hexstr = encode_structured_data(
        ...     hexstr=json.dumps(typed_data).encode("utf-8").hex()
        ... )

        >>> signed_msg_from_dict = Account.sign_message(signable_msg_from_dict, key)
        >>> signed_msg_from_str = Account.sign_message(signable_msg_from_str, key)
        >>> signed_msg_from_hexstr = Account.sign_message(signable_msg_from_hexstr, key)

        >>> signed_msg_from_dict == signed_msg_from_str == signed_msg_from_hexstr
        True
        >>> signed_msg_from_dict.messageHash
        HexBytes('0xbe609aee343fb3c4b28e1df9e632fca64fcfaede20f02e86244efddf30957bd2')

    .. _EIP-712: https://eips.ethereum.org/EIPS/eip-712
    """
    if isinstance(primitive, Mapping):
        validate_structured_data(primitive)
        structured_data = primitive
    else:
        message_string = to_text(primitive, hexstr=hexstr, text=text)
        structured_data = load_and_validate_structured_message(message_string)
    return SignableMessage(
        HexBytes(b"\x01"),
        hash_domain(structured_data),
        hash_eip712_message(structured_data),
    )


def encode_defunct(
    primitive: bytes = None, *, hexstr: str = None, text: str = None
) -> SignableMessage:
    r"""
    Encode a message for signing, using an old, unrecommended approach.

    Only use this method if you must have compatibility with
    :meth:`w3.eth.sign() <web3.eth.Eth.sign>`.

    EIP-191 defines this as "version ``E``".

    .. NOTE: This standard includes the number of bytes in the message as a part of
        the header. Awkwardly, the number of bytes in the message is encoded in
        decimal ascii. So if the message is 'abcde', then the length is encoded
        as the ascii character '5'. This is one of the reasons that this message
        format is not preferred. There is ambiguity when the message '00' is
        encoded, for example.

    Supply exactly one of the three arguments: bytes, a hex string, or a unicode string.

    :param primitive: the binary message to be signed
    :type primitive: bytes or int
    :param str hexstr: the message encoded as hex
    :param str text: the message as a series of unicode characters (a normal Py3 str)
    :returns: The EIP-191 encoded message, ready for signing

    .. doctest:: python

        >>> from eth_account.messages import encode_defunct
        >>> from eth_utils.curried import to_hex, to_bytes

        >>> message_text = "I♥SF"
        >>> encode_defunct(text=message_text)
        SignableMessage(version=b'E',
                        header=b'thereum Signed Message:\n6',
                        body=b'I\xe2\x99\xa5SF')

        These four also produce the same hash:
        >>> encode_defunct(to_bytes(text=message_text))
        SignableMessage(version=b'E',
                        header=b'thereum Signed Message:\n6',
                        body=b'I\xe2\x99\xa5SF')

        >>> encode_defunct(bytes(message_text, encoding='utf-8'))
        SignableMessage(version=b'E',
                        header=b'thereum Signed Message:\n6',
                        body=b'I\xe2\x99\xa5SF')

        >>> to_hex(text=message_text)
        '0x49e299a55346'
        >>> encode_defunct(hexstr='0x49e299a55346')
        SignableMessage(version=b'E',
                        header=b'thereum Signed Message:\n6',
                        body=b'I\xe2\x99\xa5SF')

        >>> encode_defunct(0x49e299a55346)
        SignableMessage(version=b'E',
                        header=b'thereum Signed Message:\n6',
                        body=b'I\xe2\x99\xa5SF')
    """
    message_bytes = to_bytes(primitive, hexstr=hexstr, text=text)
    msg_length = str(len(message_bytes)).encode("utf-8")

    # Encoding version E defined by EIP-191
    return SignableMessage(
        b"E",
        b"thereum Signed Message:\n" + msg_length,
        message_bytes,
    )


def defunct_hash_message(
    primitive: bytes = None, *, hexstr: str = None, text: str = None
) -> HexBytes:
    """
    Convert the provided message into a message hash, to be signed.

    .. CAUTION:: Intended for use with the deprecated
        :meth:`eth_account.account.Account.signHash`.
        This is for backwards compatibility only. All new implementations
        should use :meth:`encode_defunct` instead.

    :param primitive: the binary message to be signed
    :type primitive: bytes or int
    :param str hexstr: the message encoded as hex
    :param str text: the message as a series of unicode characters (a normal Py3 str)
    :returns: The hash of the message, after adding the prefix
    """
    signable = encode_defunct(primitive, hexstr=hexstr, text=text)
    hashed = _hash_eip191_message(signable)
    return HexBytes(hashed)
